manta-trusted-setup/src/groth16/ppot/serialization.rs

```diff
@@ -0,0 +1,964 @@
+    fn get_size(&self, compression: Compressed) -> usize {
+        match compression {
+            Compressed::No => {
+                if self.is_g1_type() {
+                    64
+                } else {
+                    128
+                }
+            }
+            Compressed::Yes => {
+                if self.is_g1_type() {
+                    32
+                } else {
+                    64
+                }
+            }
+        }
+    }
+
+    /// The number of powers of elements of this type in an accumulator.
+    fn num_powers<S>(&self) -> usize
+    where
+        S: Size,
+    {
+        match self {
+            ElementType::BetaG2 => 1,
+            ElementType::TauG1 => S::G1_POWERS,
+            _ => S::G2_POWERS,
+        }
+    }
+
+    /// Element is a point on the G1 curve
+    fn is_g1_type(&self) -> bool {
+        !matches!(self, ElementType::TauG2 | ElementType::BetaG2)
+    }
+}
+
+/// Reads appropriate number of elements of `element_type` for an accumulator of given `Size` from PPoT challenge file.
+#[inline]
+pub fn read_g1_powers<S>(
+    readable_map: &Mmap,
+    element: ElementType,
+    compression: Compressed,
+) -> Result<Vec<G1Affine>, PointDeserializeError>
+where
+    S: Size,
+{
+    let size = element.num_powers::<S>();
+    let mut powers = Vec::new();
+    let mut start_position = calculate_mmap_position(0, element, compression);
+    let mut end_position = start_position + element.get_size(compression);
+    for _ in 0..size {
+        let mut reader = readable_map
+            .get(start_position..end_position)
+            .expect("cannot read point data from file");
+
+        if element.is_g1_type() {
+            let point = match compression {
+                Compressed::No => {
+                    <PpotCeremony as Deserializer<G1Affine, G1>>::deserialize_unchecked(
+                        &mut reader,
+                    )?
+                }
+                Compressed::Yes => {
+                    <PpotCeremony as Deserializer<G1Affine, G1>>::deserialize_compressed(
+                        &mut reader,
+                    )?
+                }
+            };
+            // point will be checked below
+            powers.push(point);
+        } else {
+            panic!("Expected G1 curve points")
+        }
+        start_position = end_position;
+        end_position += element.get_size(compression);
+    }
+
+    // Do curve point checks in parallel
+    match compression {
+        Compressed::No => cfg_iter!(powers).for_each(|g| curve_point_checks(g).unwrap()),
+        Compressed::Yes => cfg_iter!(powers).for_each(|g| {
+            if !g.is_in_correct_subgroup_assuming_on_curve() {
+                panic!() // This should actually just return the NotInSubgroup error
+            }
+        }),
+    }
+    Ok(powers)
+}
+
+/// Reads `size` many elements of `element_type` from PPoT challenge file.
+#[inline]
+pub fn read_g2_powers<S>(
```

Comment:
Why not a macro that generates both `read_g1_powers` and `read_g2_powers`?

Reply:
Yeah that might be worth doing